File: internal/model/trades.go
package model

import (
	"math"
	"sort"
	"strings"

	"github.com/ProfessorBagholder/Bagholder/internal/py"
	"github.com/ProfessorBagholder/Bagholder/internal/store"
	"github.com/ProfessorBagholder/Bagholder/internal/symbols"
)

func slimSlice(s *Slice) SlimSlice {
	flags := s.Flags
	if flags == nil {
		flags = []string{}
	}
	return SlimSlice{
		Key:            sliceMemberKey(s),
		Qty:            s.Quantity,
		Entry:          s.EntryPrice,
		Exit:           s.ExitPrice,
		EntryDate:      s.EntryDate,
		ExitDate:       s.ExitDate,
		Pnl:            s.Pnl,
		PnlCad:         s.PnlCad,
		Fees:           s.Commission,
		BuyActivityID:  s.BuyActivityID,
		SellActivityID: s.SellActivityID,
		Flags:          flags,
	}
}

func fillRow(a *Act) Fill {
	when := a.OccurredAt
	if when == "" {
		when = a.TransactionDate
	}
	day, clock := whenParts(when)
	side := store.TradeSide(a)
	qty := math.Abs(a.Quantity)
	if side == "SELL" {
		qty = -qty
	}
	if day == "" {
		day = a.TransactionDate
	}
	flags := a.Flags
	if flags == nil {
		flags = []string{}
	}
	return Fill{
		ID:       a.ID,
		When:     when,
		Date:     day,
		Time:     clock,
		Side:     side,
		Sub:      a.ActivitySubType,
		Qty:      qty,
		Price:    a.UnitPrice,
		Amount:   a.NetCashAmount,
		Fees:     a.Commission,
		Currency: a.Currency,
		Flags:    flags,
	}
}

func sliceLess(a, b *Slice) bool {
	if a.ExitDate != b.ExitDate {
		return a.ExitDate < b.ExitDate
	}
	if a.EntryDate != b.EntryDate {
		return a.EntryDate < b.EntryDate
	}
	return sliceMemberKey(a) < sliceMemberKey(b)
}

func journalOf(journal map[string]store.JournalEntry, id string) (string, string, []string) {
	e, ok := journal[id]
	if !ok {
		return "", "", []string{}
	}
	tags := append([]string{}, e.Tags...)
	if tags == nil {
		tags = []string{}
	}
	return e.Grade, e.Thesis, tags
}

func collapseTrade(gid string, members []*Slice, locked bool, status string, actsByID map[string]*Act, securities *Securities, journal map[string]store.JournalEntry) *Trade {
	slices := append([]*Slice{}, members...)
	sort.SliceStable(slices, func(i, j int) bool { return sliceLess(slices[i], slices[j]) })
	t0 := slices[0]
	var qty, entryNotional, exitNotional, pnl, pnlCad, fees, feesCad float64
	entryDate, exitDate := slices[0].EntryDate, slices[0].ExitDate
	entryWhen, exitWhen := "", ""
	for i, s := range slices {
		qty += s.Quantity
		entryNotional += s.EntryPrice * s.Quantity
		exitNotional += s.ExitPrice * s.Quantity
		pnl += s.Pnl
		pnlCad += s.PnlCad
		fees += s.Commission
		if s.HasFeesCad {
			feesCad += s.FeesCad
		} else {
			feesCad += s.Commission
		}
		if s.EntryDate < entryDate {
			entryDate = s.EntryDate
		}
		if s.ExitDate > exitDate {
			exitDate = s.ExitDate
		}
		ew := s.EntryWhen
		if ew == "" {
			ew = s.EntryDate
		}
		xw := s.ExitWhen
		if xw == "" {
			xw = s.ExitDate
		}
		if i == 0 || ew < entryWhen {
			entryWhen = ew
		}
		if i == 0 || xw > exitWhen {
			exitWhen = xw
		}
	}
	mult := multiplier(t0.Symbol)
	entry := 0.0
	if qty != 0 {
		entry = entryNotional / qty
	}
	exitPx := t0.ExitPrice
	if qty != 0 {
		exitPx = exitNotional / qty
	}
	basis := math.Abs(entry * qty * mult)
	secID := ""
	for _, s := range slices {
		if s.SecurityID != "" {
			secID = s.SecurityID
			break
		}
	}
	var ids []string
	for _, s := range slices {
		for _, k := range []string{s.BuyActivityID, s.SellActivityID} {
			if k != "" && !py.Contains(ids, k) {
				ids = append(ids, k)
			}
		}
	}
	fills := []Fill{}
	for _, i := range ids {
		if a, ok := actsByID[i]; ok {
			fills = append(fills, fillRow(a))
		}
	}
	openedIDs := map[string]bool{}
	closedIDs := map[string]bool{}
	for _, s := range slices {
		openedIDs[s.BuyActivityID] = true
		closedIDs[s.SellActivityID] = true
	}
	for i := range fills {
		f := &fills[i]
		opened, closed := openedIDs[f.ID], closedIDs[f.ID]
		side := "SELL"
		if f.Side == "BUY" {
			side = "BUY"
		}
		if t0.Kind != "Options" {
			f.Sub = side
			if opened && closed {
				f.Sub = side + " (close + open)"
			}
		} else if closed && !opened {
			f.Sub = side + " TO CLOSE"
		} else if opened && !closed {
			f.Sub = side + " TO OPEN"
		} else if opened && closed {
			f.Sub = side + " (close + open)"
		}
	}
	sort.SliceStable(fills, func(i, j int) bool { return fills[i].When > fills[j].When })
	openSide := "SELL"
	if t0.OpenDirection == "LONG" {
		openSide = "BUY"
	}
	opens, closes := 0, 0
	for _, f := range fills {
		if f.Side == openSide {
			opens++
		} else {
			closes++
		}
	}
	flagSet := map[string]bool{}
	flags := []string{}
	for _, s := range slices {
		for _, fl := range s.Flags {
			if !flagSet[fl] {
				flagSet[fl] = true
				flags = append(flags, fl)
			}
		}
	}
	sort.Strings(flags)
	grade, thesis, tags := journalOf(journal, gid)
	name := t0.Name
	if name == "" {
		name = t0.Symbol
	}
	exchange := "Crypto"
	if t0.Kind != "Crypto" {
		exchange = securities.Exchange(secID)
	}
	side := "COVER"
	if t0.OpenDirection == "LONG" {
		side = "SELL"
	}
	var pnlPct *float64
	if basis > 0 {
		pnlPct = py.Ptr(pnl / basis)
	}
	legs := make([]SlimSlice, 0, len(slices))
	for _, s := range slices {
		legs = append(legs, slimSlice(s))
	}
	return &Trade{TradeCore: TradeCore{
		ID:            gid,
		Status:        status,
		Locked:        locked,
		Symbol:        t0.Symbol,
		Underlying:    underlying(t0.Symbol),
		Name:          securities.Name(secID, name),
		Exchange:      exchange,
		Kind:          t0.Kind,
		Currency:      t0.Currency,
		Account:       t0.Account,
		AccountID:     t0.AccountID,
		SecurityID:    secID,
		Side:          side,
		OpenDirection: t0.OpenDirection,
		Qty:           qty,
		Mult:          mult,
		Entry:         entry,
		Exit:          exitPx,
		EntryDate:     entryDate,
		ExitDate:      exitDate,
		EntryWhen:     entryWhen,
		ExitWhen:      exitWhen,
		HoldDays:      daysBetween(entryDate, exitDate),
		Pnl:           pnl,
		PnlCad:        pnlCad,
		Fees:          fees,
		FeesCad:       feesCad,
		PnlPct:        pnlPct,
		LegCount:      len(slices),
		Opened:        Summary{Qty: qty, Avg: entry, Fills: opens},
		Closed:        Summary{Qty: qty, Avg: exitPx, Fills: closes},
		NetCash:       pnl,
		Flags:         flags,
		Grade:         grade,
		Thesis:        thesis,
		Tags:          tags,
	}, Legs: legs, Fills: fills}
}

func BuildTrades(closed []*Slice, openLots []*Lot, saved []store.TradeGroup, actsByID map[string]*Act, securities *Securities, journal map[string]store.JournalEntry) []*Trade {
	byKey := map[string]*Slice{}
	for _, s := range closed {
		byKey[sliceMemberKey(s)] = s
	}
	used := map[string]bool{}
	type group struct {
		id      string
		members []*Slice
		locked  bool
	}
	var groups []group
	for _, rec := range saved {
		var members []*Slice
		for _, k := range rec.Members {
			s := byKey[k]
			if s == nil {
				continue
			}
			mk := sliceMemberKey(s)
			if !used[mk] {
				members = append(members, s)
				used[mk] = true
			}
		}
		if len(members) > 0 {
			id := rec.ID
			if id == "" {
				keys := make([]string, 0, len(members))
				for _, m := range members {
					keys = append(keys, sliceMemberKey(m))
				}
				id = GroupIDForKeys(keys)
			}
			groups = append(groups, group{id, members, true})
		}
	}
	byRT := map[string][]*Slice{}
	var order []string
	for _, s := range closed {
		if used[sliceMemberKey(s)] {
			continue
		}
		rt := s.RT
		if rt == "" {
			rt = "rt:" + sliceMemberKey(s)
		}
		if _, ok := byRT[rt]; !ok {
			order = append(order, rt)
		}
		byRT[rt] = append(byRT[rt], s)
	}
	for _, rt := range order {
		groups = append(groups, group{rt, byRT[rt], false})
	}
	trades := make([]*Trade, 0, len(groups))
	for _, g := range groups {
		trades = append(trades, collapseTrade(g.id, g.members, g.locked, "closed", actsByID, securities, journal))
	}
	sort.SliceStable(trades, func(i, j int) bool {
		a, b := trades[i], trades[j]
		if a.ExitDate != b.ExitDate {
			return a.ExitDate > b.ExitDate
		}
		return a.ID > b.ID
	})
	return trades
}

func LastFillPrices(activities []*Act) map[string]LastPrice {
	acts := append([]*Act{}, activities...)
	sort.SliceStable(acts, func(i, j int) bool {
		if acts[i].TransactionDate != acts[j].TransactionDate {
			return acts[i].TransactionDate < acts[j].TransactionDate
		}
		return acts[i].OccurredAt < acts[j].OccurredAt
	})
	out := map[string]LastPrice{}
	for _, a := range acts {
		if a.Category != "trade" && a.Category != "option_event" {
			continue
		}
		if a.UnitPrice > 0 && a.Symbol != "" {
			out[a.Symbol] = LastPrice{Price: a.UnitPrice, Date: a.TransactionDate}
		}
	}
	return out
}

func UnderTicker[T any](mapping map[string]T, symbol string) (T, bool) {
	if v, ok := mapping[symbol]; ok {
		return v, true
	}
	v, ok := mapping[symbols.TMXSymbol(symbol)]
	return v, ok
}

func QuoteFits(quote *store.Quote, kind string) bool {
	if quote == nil || quote.Source == "" {
		return true
	}
	source := quote.Source
	if kind == "Crypto" {
		return source == "coinbase"
	}
	if kind == "Options" {
		return source == "cboe_options"
	}
	return source != "coinbase" && source != "cboe_options"
}

type balKey struct{ account, security string }

func BuildPositions(openLots []*Lot, lastPrices map[string]LastPrice, balances []store.Balance, accounts []store.Account, securities *Securities, journal map[string]store.JournalEntry, today string, quotes map[string]store.Quote, actsByID map[string]*Act) []*Position {
	nickIDs := map[string][]string{}
	for _, acc := range accounts {
		nick := normAccountName(firstNonEmpty(acc.Nickname, acc.UnifiedAccountType, acc.Type))
		if !py.Contains(nickIDs[nick], acc.ID) {
			nickIDs[nick] = append(nickIDs[nick], acc.ID)
		}
	}
	bal := map[balKey]float64{}
	for _, b := range balances {
		k := balKey{b.AccountID, b.SecurityID}
		bal[k] = bal[k] + py.Deref(b.Quantity, 0)
	}
	type gkey struct{ symbol, account, currency, direction string }
	groups := map[gkey][]*Lot{}
	var order []gkey
	for _, lot := range openLots {
		k := gkey{lot.Symbol, lot.AccountType, lot.Currency, lot.Direction}
		if _, ok := groups[k]; !ok {
			order = append(order, k)
		}
		groups[k] = append(groups[k], lot)
	}
	rows := []*Position{}
	for _, k := range order {
		lots := append([]*Lot{}, groups[k]...)
		sort.SliceStable(lots, func(i, j int) bool {
			if lots[i].Date != lots[j].Date {
				return lots[i].Date < lots[j].Date
			}
			return lots[i].When < lots[j].When
		})
		symbol, account, currency, direction := k.symbol, k.account, k.currency, k.direction
		mult := multiplier(symbol)
		qty := 0.0
		for _, l := range lots {
			qty += l.Qty
		}
		if qty <= 1e-9 {
			continue
		}
		cost, fees := 0.0, 0.0
		for _, l := range lots {
			cost += l.Qty * l.Price * mult
			fees += l.Commission
		}
		secID := ""
		for _, l := range lots {
			if l.SecurityID != "" {
				secID = l.SecurityID
				break
			}
		}
		last, hasLast := lastPrices[symbol]
		lastPx := 0.0
		lastAt := ""
		if hasLast {
			lastPx = last.Price
			lastAt = last.Date
		} else if qty != 0 {
			lastPx = cost / (qty * mult)
		}
		priceSource := "fill"
		var quote *store.Quote
		if q, ok := UnderTicker(quotes, symbol); ok {
			qc := q
			quote = &qc
			if !QuoteFits(quote, lots[0].Kind) {
				quote = nil
			}
		}
		if quote != nil && quote.Price != nil && *quote.Price != 0 {
			lastPx = *quote.Price
			lastAt = quote.FetchedAt
			priceSource = "quote"
		}
		mv := qty * lastPx * mult
		unreal := mv - cost
		if direction != "LONG" {
			unreal = cost - mv
		}
		held := 0.0
		for _, l := range lots {
			held += l.Qty * float64(daysBetween(l.Date, today))
		}
		var wsQty *float64
		if secID != "" {
			if ids, ok := nickIDs[account]; ok {
				total := 0.0
				found := false
				for _, aid := range ids {
					if v, ok := bal[balKey{aid, secID}]; ok {
						total += v
						found = true
					}
				}
				if found {
					wsQty = py.Ptr(total)
				}
			}
		}
		legacyPID := "pos:" + strings.Join([]string{account, symbol, currency}, "|")
		pid := lots[0].RT
		if pid == "" {
			pid = legacyPID
		}
		jid := pid
		if _, ok := journal[pid]; !ok {
			jid = legacyPID
		}
		grade, thesis, tags := journalOf(journal, jid)
		var priceChange, percentChange, dayChange *float64
		if quote != nil {
			priceChange = copyPtr(quote.PriceChange)
			percentChange = copyPtr(quote.PercentChange)
		}
		if priceChange != nil {
			sign := 1.0
			if direction == "SHORT" {
				sign = -1
			}
			dayChange = py.Ptr(qty * *priceChange * mult * sign)
		}
		fills := []Fill{}
		for _, l := range lots {
			if a, ok := actsByID[l.ActivityID]; ok {
				fills = append(fills, fillRow(a))
			}
		}
		sort.SliceStable(fills, func(i, j int) bool { return fills[i].When > fills[j].When })
		plots := make([]PositionLot, 0, len(lots))
		for _, l := range lots {
			flags := l.Flags
			if flags == nil {
				flags = []string{}
			}
			plots = append(plots, PositionLot{Opened: l.Date, Qty: l.Qty, Price: l.Price, Basis: l.Qty * l.Price * mult, Held: daysBetween(l.Date, today), Flags: flags, ActivityID: l.ActivityID})
		}
		name := lots[0].Name
		if name == "" {
			name = symbol
		}
		exchange := "Crypto"
		if lots[0].Kind != "Crypto" {
			exchange = securities.Exchange(secID)
		}
		var unrealPct *float64
		if cost != 0 {
			unrealPct = py.Ptr(unreal / cost)
		}
		heldDays := 0
		if qty != 0 {
			heldDays = py.RoundInt(held / qty)
		}
		avg := 0.0
		if qty != 0 {
			avg = cost / (qty * mult)
		}
		var rt *string
		if lots[0].RT != "" {
			r := lots[0].RT
			rt = &r
		}
		rows = append(rows, &Position{PositionCore: PositionCore{
			ID:            pid,
			Symbol:        symbol,
			Underlying:    underlying(symbol),
			Name:          securities.Name(secID, name),
			Exchange:      exchange,
			Kind:          lots[0].Kind,
			Account:       account,
			AccountID:     lots[0].AccountID,
			Currency:      currency,
			SecurityID:    secID,
			Short:         direction == "SHORT",
			Qty:           qty,
			Mult:          mult,
			Avg:           avg,
			Cost:          cost,
			Fees:          fees,
			Last:          lastPx,
			LastAt:        lastAt,
			PriceSource:   priceSource,
			PriceChange:   priceChange,
			PercentChange: percentChange,
			DayChange:     dayChange,
			MV:            mv,
			Unreal:        unreal,
			UnrealPct:     unrealPct,
			Held:          heldDays,
			Opened:        lots[0].Date,
			WsQty:         wsQty,
			RT:            rt,
			Lots:          plots,
			Grade:         grade,
			Thesis:        thesis,
			Tags:          tags,
		}, Fills: fills})
	}
	book := 0.0
	for _, r := range rows {
		book += math.Abs(r.Cost)
	}
	for _, r := range rows {
		if book != 0 {
			r.Alloc = math.Abs(r.Cost) / book
		}
	}
	sort.SliceStable(rows, func(i, j int) bool { return rows[i].Alloc > rows[j].Alloc })
	return rows
}

func copyPtr(p *float64) *float64 {
	if p == nil {
		return nil
	}
	v := *p
	return &v
}

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

func BuildCashflow(activities []*Act, securities *Securities, fx map[string]float64) []CashRow {
	rows := []CashRow{}
	for _, a := range activities {
		cat := a.Category
		raw := compact(a.RawType)
		at := compact(a.ActivityType)
		cash := a.NetCashAmount
		kind := ""
		switch {
		case cat == "dividend":
			kind = "Dividend"
		case cat == "interest":
			kind = "Interest"
		case raw == "WITHHOLDINGTAX" || at == "WITHHOLDINGTAX":
			kind = "Withholding tax"
		case raw == "INTERESTCHARGE" || at == "INTERESTCHARGE":
			kind = "Interest charge"
		default:
			continue
		}
		if math.Abs(cash) < EPS {
			continue
		}
		when := a.OccurredAt
		if when == "" {
			when = a.TransactionDate
		}
		day, clock := whenParts(when)
		symbol := py.Strip(a.Symbol)
		if symbol == "" && (kind == "Interest" || kind == "Interest charge") {
			symbol = "Cash"
		}
		nameFallback := ""
		if a.Name != symbol {
			nameFallback = a.Name
		}
		date := a.TransactionDate
		if date == "" {
			date = day
		}
		shown := symbol
		if shown == "" {
			shown = "—"
		}
		account := normAccountName(a.AccountType)
		if account == "" {
			account = a.AccountID
		}
		var qty, per *float64
		if a.Quantity != 0 {
			qty = py.Ptr(a.Quantity)
		}
		if a.UnitPrice != 0 {
			per = py.Ptr(a.UnitPrice)
		}
		ccy := a.Currency
		if ccy == "" {
			ccy = "CAD"
		}
		rows = append(rows, CashRow{
			ID:        a.ID,
			Date:      date,
			Time:      clock,
			Symbol:    shown,
			Name:      securities.Name(a.SecurityID, nameFallback),
			Kind:      kind,
			Account:   account,
			AccountID: a.AccountID,
			Qty:       qty,
			Per:       per,
			Amount:    cash,
			Currency:  ccy,
			AmountCad: toCad(fx, cash, a.Currency, a.TransactionDate),
		})
	}
	sort.SliceStable(rows, func(i, j int) bool {
		if rows[i].Date != rows[j].Date {
			return rows[i].Date > rows[j].Date
		}
		return rows[i].ID > rows[j].ID
	})
	return rows
}

File: rust/crates/model/src/trades.rs
//! Round trips: `build_trades` and `collapse_trade`.
//!
//! A trade is a position going from flat to open and back to flat. Partial
//! exits are legs of the same trade, and the id is stable from the first fill
//! so a journal entry survives later exits.

use std::collections::{HashMap, HashSet};
use std::sync::Arc;

use crate::activity::{Activity, Direction, Flag, Kind, Side};
use crate::book::Book;
use crate::clock::when_parts;
use crate::dates::days_between;
use crate::fifo::{trade_side, Slice};
use crate::input::{Journal, LastFill, TradeGroup};
use crate::symbols::{option_multiplier, underlying_symbol};
use crate::value::{fmt8, FSum};
use crate::wire::{ExitSide, Fill, Leg, Tally, Trade};

/// What a saved group names its members by.
pub fn slice_member_key(t: &Slice) -> String {
    if !t.buy_activity_id.is_empty() && !t.sell_activity_id.is_empty() {
        return [t.buy_activity_id.clone(), t.sell_activity_id.clone(), fmt8(t.quantity)].join("|");
    }
    t.id.clone()
}

/// The legacy page's group id: FNV-1a over the sorted member keys.
pub fn group_id_for_keys(keys: &[String]) -> String {
    let mut sorted: Vec<&String> = keys.iter().collect();
    sorted.sort();
    let joined = sorted.iter().map(|s| s.as_str()).collect::<Vec<_>>().join("\n");
    let mut h: u32 = 2166136261;
    for ch in joined.chars() {
        h ^= ch as u32;
        h = h.wrapping_mul(16777619);
    }
    format!("g_{:x}_{}", h, keys.len())
}

fn leg(s: &Slice) -> Leg {
    Leg {
        key: slice_member_key(s),
        qty: s.quantity,
        entry: s.entry_price,
        exit: s.exit_price,
        entry_date: s.entry_date.clone(),
        exit_date: s.exit_date.clone(),
        pnl: s.pnl,
        pnl_cad: s.pnl_cad,
        fees: s.commission,
        buy_activity_id: s.buy_activity_id.clone(),
        sell_activity_id: s.sell_activity_id.clone(),
        flags: s.flags.clone(),
    }
}

/// One broker fill as the page prints it.
pub fn fill_row(a: &Activity) -> Fill {
    let (day, clock) = when_parts(a.when());
    let side = trade_side(a);
    let qty = a.quantity.abs();
    Fill {
        id: a.id.clone(),
        when: a.when().to_string(),
        date: if day.is_empty() { a.transaction_date.clone() } else { day },
        time: clock,
        side,
        sub: a.activity_sub_type.clone(),
        qty: if side == Some(Side::Sell) { -qty } else { qty },
        price: a.unit_price,
        amount: a.net_cash_amount,
        fees: a.commission,
        currency: a.currency.clone(),
        flags: a.flags.clone(),
    }
}

/// One group of slices as a single round trip.
pub fn collapse_trade(gid: &str, slices: &[Slice], locked: bool, book: &Book, journal: &Journal) -> Trade {
    let mut slices: Vec<&Slice> = slices.iter().collect();
    slices.sort_by_cached_key(|s| (s.exit_date.clone(), s.entry_date.clone(), slice_member_key(s)));
    let t0 = slices[0];

    let qty: f64 = slices.iter().map(|s| s.quantity).fsum();
    let entry_notional: f64 = slices.iter().map(|s| s.entry_price * s.quantity).fsum();
    let exit_notional: f64 = slices.iter().map(|s| s.exit_price * s.quantity).fsum();
    let pnl: f64 = slices.iter().map(|s| s.pnl).fsum();
    let pnl_cad: f64 = slices.iter().map(|s| s.pnl_cad).fsum();
    let fees: f64 = slices.iter().map(|s| s.commission).fsum();
    let fees_cad: f64 = slices.iter().map(|s| s.fees_cad.unwrap_or(s.commission)).fsum();
    let entry_date = slices.iter().map(|s| s.entry_date.clone()).min().unwrap_or_default();
    let exit_date = slices.iter().map(|s| s.exit_date.clone()).max().unwrap_or_default();
    let or_day = |when: &String, day: &String| if when.is_empty() { day.clone() } else { when.clone() };
    let entry_when = slices.iter().map(|s| or_day(&s.entry_when, &s.entry_date)).min().unwrap_or_default();
    let exit_when = slices.iter().map(|s| or_day(&s.exit_when, &s.exit_date)).max().unwrap_or_default();

    let mult = option_multiplier(&t0.symbol);
    let entry = if qty != 0.0 { entry_notional / qty } else { 0.0 };
    let exit = if qty != 0.0 { exit_notional / qty } else { t0.exit_price };
    let basis = (entry * qty * mult).abs();
    let security_id = slices.iter().map(|s| &s.security_id).find(|s| !s.is_empty()).cloned().unwrap_or_default();

    let mut ids: Vec<&String> = Vec::new();
    for s in &slices {
        for id in [&s.buy_activity_id, &s.sell_activity_id] {
            if !id.is_empty() && !ids.contains(&id) {
                ids.push(id);
            }
        }
    }
    let mut fills: Vec<Fill> = ids.iter().filter_map(|id| book.activity(id)).map(fill_row).collect();

    // Label each fill by what it did in this trade, not by the broker's order
    // type: the open/close types are option language, while shares and crypto
    // are simply bought or sold.
    let opened_ids: HashSet<&String> = slices.iter().map(|s| &s.buy_activity_id).collect();
    let closed_ids: HashSet<&String> = slices.iter().map(|s| &s.sell_activity_id).collect();
    for f in fills.iter_mut() {
        let (opened, closed) = (opened_ids.contains(&f.id), closed_ids.contains(&f.id));
        let side = if f.side == Some(Side::Buy) { "BUY" } else { "SELL" };
        f.sub = match (t0.kind, opened, closed) {
            (Kind::Options, false, true) => format!("{} TO CLOSE", side),
            (Kind::Options, true, false) => format!("{} TO OPEN", side),
            (Kind::Options, false, false) => f.sub.clone(),
            (_, true, true) => format!("{} (close + open)", side),
            _ => side.to_string(),
        };
    }
    fills.sort_by(|a, b| b.when.cmp(&a.when));

    let opens = fills.iter().filter(|f| f.side == Some(t0.open_direction.opened_by())).count();
    let closes = fills.len() - opens;

    let mut flags: Vec<Flag> = slices.iter().flat_map(|s| s.flags.iter().cloned()).collect();
    flags.sort();
    flags.dedup();

    let said = journal.get(gid).cloned().unwrap_or_default();
    Trade {
        id: gid.to_string(),
        status: "closed",
        locked,
        symbol: t0.symbol.clone(),
        underlying: underlying_symbol(&t0.symbol),
        name: book.securities.name(&security_id, if t0.name.is_empty() { &t0.symbol } else { &t0.name }),
        exchange: if t0.kind == Kind::Crypto { "Crypto".to_string() } else { book.securities.exchange(&security_id) },
        kind: t0.kind,
        currency: t0.currency.clone(),
        account: t0.account.clone(),
        account_id: t0.account_id.clone(),
        security_id,
        side: if t0.open_direction == Direction::Long { ExitSide::Sell } else { ExitSide::Cover },
        open_direction: t0.open_direction,
        qty,
        mult: mult as i64,
        entry,
        exit,
        hold_days: days_between(&entry_date, &exit_date),
        entry_date,
        exit_date,
        entry_when,
        exit_when,
        pnl,
        pnl_cad,
        fees,
        fees_cad,
        pnl_pct: (basis > 0.0).then(|| pnl / basis),
        legs: Some(Arc::new(slices.iter().map(|s| leg(s)).collect())),
        leg_count: slices.len(),
        opened: Tally { qty, avg: entry, fills: opens },
        closed: Tally { qty, avg: exit, fills: closes },
        fills: Some(Arc::new(fills)),
        net_cash: pnl,
        flags,
        grade: said.grade,
        thesis: said.thesis,
        tags: said.tags,
    }
}

/// The saved manual groups first, then whatever is left grouped by round trip.
pub fn build_trades(closed: &[Slice], saved_groups: &[TradeGroup], book: &Book, journal: &Journal) -> Vec<Trade> {
    let by_key: HashMap<String, &Slice> = closed.iter().map(|s| (slice_member_key(s), s)).collect();
    let mut used: HashSet<String> = HashSet::new();
    let mut groups: Vec<(String, Vec<Slice>, bool)> = Vec::new();

    for saved in saved_groups {
        let mut members: Vec<Slice> = Vec::new();
        for key in &saved.members {
            if let Some(s) = by_key.get(key) {
                if used.insert(key.clone()) {
                    members.push((*s).clone());
                }
            }
        }
        if !members.is_empty() {
            let gid = if saved.id.is_empty() { group_id_for_keys(&members.iter().map(slice_member_key).collect::<Vec<_>>()) } else { saved.id.clone() };
            groups.push((gid, members, true));
        }
    }

    let mut by_rt: HashMap<String, Vec<Slice>> = HashMap::new();
    let mut order: Vec<String> = Vec::new();
    for s in closed {
        let key = slice_member_key(s);
        if used.contains(&key) {
            continue;
        }
        let mut rt = s.rt.clone().unwrap_or_else(|| format!("rt:{}", key));
        // a deposited (unknown-basis) leg stands as its own unscoreable trade
        if s.flags.contains(&Flag::BasisUnknown) {
            rt.push_str("|nobasis");
        }
        if !by_rt.contains_key(&rt) {
            order.push(rt.clone());
        }
        by_rt.entry(rt).or_default().push(s.clone());
    }
    for rt in order {
        let members = by_rt.remove(&rt).unwrap();
        groups.push((rt, members, false));
    }

    let mut trades: Vec<Trade> = groups.iter().map(|(gid, members, locked)| collapse_trade(gid, members, *locked, book, journal)).collect();
    trades.sort_by(|a, b| (&b.exit_date, &b.id).cmp(&(&a.exit_date, &a.id)));
    trades
}

/// Symbol -> the newest fill that carried a price.
pub fn last_fill_prices(activities: &[Activity]) -> HashMap<String, LastFill> {
    let mut ordered: Vec<&Activity> = activities.iter().filter(|a| a.category.is_fill() && a.unit_price > 0.0 && !a.symbol.is_empty()).collect();
    ordered.sort_by(|a, b| (&a.transaction_date, &a.occurred_at).cmp(&(&b.transaction_date, &b.occurred_at)));
    ordered.into_iter().map(|a| (a.symbol.clone(), LastFill { price: a.unit_price, date: a.transaction_date.clone() })).collect()
}
